src/Features/LanguageServer/Protocol/Handler/Configuration/DidChangeConfigurationNotificationHandler.cs

```diff
@@ -0,0 +1,212 @@
+            .Add(LanguageNames.CSharp, "csharp")
+            .Add(LanguageNames.VisualBasic, "visual_basic");
+
+        public static readonly ImmutableArray<string> SupportedLanguages = ImmutableArray.Create(LanguageNames.CSharp, LanguageNames.VisualBasic);
+
+        public DidChangeConfigurationNotificationHandler(
+            ILspLogger logger,
+            IGlobalOptionService globalOptionService,
+            IClientLanguageServerManager clientLanguageServerManager,
+            IAsynchronousOperationListener asynchronousOperationListener)
+        {
+            _lspLogger = logger;
+            _globalOptionService = globalOptionService;
+            _clientLanguageServerManager = clientLanguageServerManager;
+            _registrationId = Guid.NewGuid();
+            _asynchronousOperationListener = asynchronousOperationListener;
+            _configurationItems = GenerateGlobalConfigurationItems();
+        }
+
+        public bool MutatesSolutionState => true;
+
+        public bool RequiresLSPSolution => false;
+
+        public async Task HandleNotificationAsync(DidChangeConfigurationParams request, RequestContext requestContext, CancellationToken cancellationToken)
+        {
+            using var _ = _asynchronousOperationListener.BeginAsyncOperation(Methods.WorkspaceDidChangeConfigurationName);
+            await RefreshOptionsAsync(cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task RefreshOptionsAsync(CancellationToken cancellationToken)
+        {
+            var configurationsFromClient = await GetConfigurationsAsync(cancellationToken).ConfigureAwait(false);
+            if (configurationsFromClient.IsEmpty)
+            {
+                // Failed to get values from client, do nothing.
+                return;
+            }
+
+            // We always fetch VB and C# value from client if the option is IPerLanguageValuedOption.
+            RoslynDebug.Assert(configurationsFromClient.Length == SupportedOptions.Sum(option => option is IPerLanguageValuedOption ? 2 : 1));
+            var optionsToRefresh = SupportedOptions.SelectManyAsArray(option => option is IPerLanguageValuedOption
+                ? SupportedLanguages.SelectAsArray(language => (option, language))
+                : SpecializedCollections.SingletonEnumerable((option, string.Empty)));
+
+            // LSP ensures the order of result from client should match the order we sent from server.
+            for (var i = 0; i < configurationsFromClient.Length; i++)
+            {
+                var valueFromClient = configurationsFromClient[i];
+                if (valueFromClient == string.Empty)
+                {
+                    // Configuration doesn't exist in client.
+                    continue;
+                }
+
+                var (option, languageName) = optionsToRefresh[i];
+                if (option is IPerLanguageValuedOption perLanguageValuedOption)
```

Comment:
Do you need this if/else?  Already have a IPerLanguageValuedOption check inside SetOption